server/balancer.go

```diff
@@ -0,0 +1,316 @@
+		}
+	}
+
+	return nil
+}
+
+// BalanceOperator is used to do region balance.
+type BalanceOperator struct {
+	index int
+	ops   []Operator
+}
+
+func newBalanceOperator(oldLeader, newLeader, newPeer *metapb.Peer) *BalanceOperator {
+	leaderTransferOperator := newTransferLeaderOperator(oldLeader, newLeader)
+
+	addPeer := &pdpb.ChangePeer{
+		ChangeType: raftpb.ConfChangeType_AddNode.Enum(),
+		Peer:       newPeer,
+	}
+	addPeerOperator := newChangePeerOperator(addPeer)
+
+	removePeer := &pdpb.ChangePeer{
+		ChangeType: raftpb.ConfChangeType_RemoveNode.Enum(),
+		Peer:       oldLeader,
+	}
+	removePeerOperator := newChangePeerOperator(removePeer)
+
+	return &BalanceOperator{
+		index: 0,
+		ops:   []Operator{leaderTransferOperator, addPeerOperator, removePeerOperator},
+	}
+}
+
+// Check implements Operator.Check interface.
+func (bo *BalanceOperator) Check(req *pdpb.RegionHeartbeatRequest) bool {
+	if bo.index >= len(bo.ops) {
+		return true
+	}
+
+	return false
+}
+
+// Do implements Operator.Do interface.
+func (bo *BalanceOperator) Do(req *pdpb.RegionHeartbeatRequest) (bool, *pdpb.RegionHeartbeatResponse) {
+	// TODO: optimize it to return next operator response directly.
+	if bo.Check(req) {
+		return true, nil
+	}
+
+	finished, res := bo.ops[bo.index].Do(req)
+	if !finished {
+		return false, res
+	}
+
+	bo.index++
+	return false, nil
+}
+
+// ChangePeerOperator is used to do peer change.
+type ChangePeerOperator struct {
+	changePeer *pdpb.ChangePeer
+}
+
+func newChangePeerOperator(changePeer *pdpb.ChangePeer) *ChangePeerOperator {
+	return &ChangePeerOperator{
+		changePeer: changePeer,
+	}
+}
+
+// Check implements Operator.Check interface.
+func (co *ChangePeerOperator) Check(req *pdpb.RegionHeartbeatRequest) bool {
+	region := req.GetRegion()
+	if region == nil {
+		return false
+	}
+
+	leaderPeer := req.GetLeader()
+	if leaderPeer == nil {
+		return false
+	}
+
+	if co.changePeer.GetChangeType() == raftpb.ConfChangeType_AddNode {
+		if containPeer(region, leaderPeer) {
+			return true
+		}
+	} else if co.changePeer.GetChangeType() == raftpb.ConfChangeType_RemoveNode {
+		if !containPeer(region, leaderPeer) {
+			return true
+		}
+	}
+
+	return false
+}
+
+// Do implements Operator.Do interface.
+func (co *ChangePeerOperator) Do(req *pdpb.RegionHeartbeatRequest) (bool, *pdpb.RegionHeartbeatResponse) {
+	if co.Check(req) {
+		return true, nil
+	}
+
+	res := &pdpb.RegionHeartbeatResponse{
+		ChangePeer: co.changePeer,
+	}
+	return false, res
+}
+
+// TransferLeaderOperator is used to do leader transfer.
+type TransferLeaderOperator struct {
+	oldLeader *metapb.Peer
+	newLeader *metapb.Peer
+}
+
+func newTransferLeaderOperator(oldLeader, newLeader *metapb.Peer) *TransferLeaderOperator {
+	return &TransferLeaderOperator{
+		oldLeader: oldLeader,
+		newLeader: newLeader,
+	}
+}
+
+// Check implements Operator.Check interface.
+func (lto *TransferLeaderOperator) Check(req *pdpb.RegionHeartbeatRequest) bool {
+	leader := req.GetLeader()
```

Comment:
Yes, that is a question, if that, we should continue the balance step although the new leader is not what we want.